Add large runtime resource directory to lfs, and expose it to the Docker build.

cmnbroad opened the pull request:
This adds a new folder for large runtime resources to git-lfs. Unlike the large test resources, which are only accessed via a volume mount when tests are run on the Docker image, the runtime resources need to be accessible to the GATK build during the Docker build process, since they're included in the jar. AFAICT there is no `docker build` equivalent to `docker run -v`. So for now the runtime resources are git pulled into the Docker staging area, and thus onto the Docker image. We need this for @lucidtronix 's CNN branch (and possibly for @TedBrookings) if we're going to load models for resources, but longer term, we need a better solution.

cmnbroad commented:
@jamesemery @lbergelson Any alternate suggestions ?

lbergelson commented on src/test/java/org/broadinstitute/hellbender/utils/io/IOUtilsUnitTest.java:
```diff
@@ -270,5 +267,21 @@ public void testUrlEncodeDecode(final String string, final String encoded) {
 
+    @Test
+    public void testLoadLargeRuntimeResource() throws IOException {
+        final String largeResourcePath = "large/largeResourceTest.txt";
```
this path should be a constant somewhere

cmnbroad commented:
I added a check to build.gradle to verify that the large (runtime) resources are not lfs pointer files. I also added a test to verify that there is a .git folder, and non-null ToolProvider classLoader (fixes https://github.com/broadinstitute/gatk/issues/4532). To test the pointer file check, I created a branch of this branch with a temporary commit that reverts the addition of the large resources to the Docker staging area, and [submitted](https://travis-ci.org/broadinstitute/gatk/builds/354495062) it to Travis to verify that the build fails when they're not present.

droazen commented on README.md:
```diff
@@ -55,6 +55,8 @@ releases of the toolkit.
     * Gradle 3.1 or greater, needed for building the GATK. We recommend using the `./gradlew` script which will
       download and use an appropriate gradle version automatically (see examples below).
+    * [git-lfs](https://git-lfs.github.com/) 1.1.0 or greater, needed for building the GATK. Used to download the large files required to run the build and to run the test suite.
+      Run `git lfs install` after downloading, followed by `git lfs pull` from the root of your git clone to download the large files. The download is several hundred megabytes.
```
Maybe we should split into "build" and "runtime" requirements?

droazen replied:
Also, can you check whether `git lfs pull` still works?

cmnbroad replied:
I split this into separate sections for runtime and build. Updated to state that the lfs download is around 2GB. And `git lfs pull` does work.

droazen commented on README.md:
```diff
@@ -55,6 +55,8 @@ releases of the toolkit.
 * Optional, but recommended:
     * Gradle 3.1 or greater, needed for building the GATK. We recommend using the `./gradlew` script which will
       download and use an appropriate gradle version automatically (see examples below).
+    * [git-lfs](https://git-lfs.github.com/) 1.1.0 or greater, needed for building the GATK. Used to download the large files required to run the build and to run the test suite.
```
"run the build" -> "build GATK"

cmnbroad replied:
done.

droazen commented on README.md:
```diff
@@ -102,7 +102,7 @@ You can download and run pre-built versions of GATK4 from the following places:
     * This creates a zip archive in the `build/` directory with a name like `gatk-VERSION.zip` containing a complete standalone GATK distribution, including our launcher `gatk`, both the local and spark jars, and this README.    
     * You can also run GATK commands directly from the root of your git clone after running this command.
-    * Note that you *must* have a full git clone in order to build GATK. The zipped source code alone is not buildable.
+    * Note that you *must* have a full git clone in order to build GATK, incuding git-lfs files. The zipped source code alone is not buildable.
```
incuding -> including

droazen replied:
"including the git-lfs files in `src/main/resources`" ?

cmnbroad replied:
done.

droazen commented on build.gradle:
```diff
@@ -38,6 +38,41 @@ startScripts {
+task validateBuildPrerequisites {
+    if (ToolProvider.getSystemToolClassLoader() == null) {
+        throw new RuntimeException(
+                String.format(buildPrerequisitesErrorFormatString, "A (version 8) JDK must be installed."));
```
Extract required Java version as a named constant at the top

cmnbroad replied:
Done.

droazen commented on build.gradle:
```diff
@@ -38,6 +38,41 @@ startScripts {
+    if (!file(".git").isDirectory()) {
+        throw new RuntimeException(
+                String.format(buildPrerequisitesErrorFormatString,
+                        "The GATK Github repository must be cloned to run the build."));
```
Add "using `git clone`" 

Also provide a link to github tutorial

cmnbroad replied:
These are all switched to use string interpolation now. The embedded buildPrerequisites message directs the user to the gatk github repo.

droazen commented on build.gradle:
```diff
@@ -38,6 +38,41 @@ startScripts {
+                        "The GATK Github repository must be cloned to run the build."));
+    }
+    // Next, ensure that the git-lfs files required to run the build are present
+    final FileTree largeResources = fileTree(dir: "src/main/resources/large")
```
Declare `src/main/resources/large` as a named constant at the top

cmnbroad replied:
Done.

droazen commented on build.gradle:
```diff
@@ -38,6 +38,41 @@ startScripts {
+def validateLargeResourceFile(final File largeResourceFile, buildPrerequisitesErrorFormatString) {
+    largeResourceFile.withReader("UTF-8") { reader ->
+        final String firstLine = reader.readLine();
+        if (firstLine != null && firstLine.startsWith("version https://git-lfs.github.com/spec/v1")) {
```
Declare this string as a named constant somewhere

cmnbroad replied:
done.

droazen commented on build.gradle:
```diff
@@ -38,6 +38,41 @@ startScripts {
+// that the lfs files have been fetched and that the directory does not contain the git-lfs pointer files
+def validateLargeResourceFile(final File largeResourceFile, buildPrerequisitesErrorFormatString) {
+    largeResourceFile.withReader("UTF-8") { reader ->
+        final String firstLine = reader.readLine();
```
Read first X bytes instead of a full line in case the first line is very large.

cmnbroad replied:
done.

droazen commented on build_docker.sh:
```diff
@@ -98,6 +98,11 @@ if [ -n "$STAGING_DIR" ]; then
     echo "${GIT_CHECKOUT_COMMAND}"
     ${GIT_CHECKOUT_COMMAND}
+    # pull just the large runtime resources into the staging area so they are available to the build
+    # when its done as part of the Docker build so the resources can be compiled into the jar as resources
```
its -> it's

cmnbroad replied:
done.

droazen commented on build_docker.sh:
```diff
@@ -98,6 +98,11 @@ if [ -n "$STAGING_DIR" ]; then
     ${GIT_CHECKOUT_COMMAND}
+    # pull just the large runtime resources into the staging area so they are available to the build
+    # when its done as part of the Docker build so the resources can be compiled into the jar as resources
+    GIT_PULL_LARGE_COMMAND="git lfs pull --include src/main/resources/large/"
```
Might be good to mention this command in the README for users who want to build GATK without downloading all of the large test data.

cmnbroad replied:
Done.

droazen commented on build_docker.sh:
```diff
@@ -98,6 +98,11 @@ if [ -n "$STAGING_DIR" ]; then
+    # when its done as part of the Docker build so the resources can be compiled into the jar as resources
+    GIT_PULL_LARGE_COMMAND="git lfs pull --include src/main/resources/large/"
+    echo ${GIT_PULL_LARGE_COMMAND}
+    ${GIT_PULL_LARGE_COMMAND}
```
Check the exit status code here

droazen commented on src/main/java/org/broadinstitute/hellbender/utils/io/Resource.java:
```diff
@@ -9,6 +9,7 @@
 public final class Resource {
     private final String path;
     private final Class<?> relativeClass;
+    public final static String LARGE_RESOURCES_PATH = "large";
```
Add a comment here explaining what this is

cmnbroad replied:
Done.

droazen commented on src/test/java/org/broadinstitute/hellbender/utils/io/IOUtilsUnitTest.java:
```diff
@@ -270,5 +267,20 @@ public void testUrlEncodeDecode(final String string, final String encoded) {
+    @Test
+    public void testLoadLargeRuntimeResource() throws IOException {
+        final Resource largeResource = new Resource(Resource.LARGE_RESOURCES_PATH + "/largeResourceTest.txt", null);
+        final File resourceFile = IOUtils.writeTempResource(largeResource);
```
Might want to extract a simple utility method for loading a resource and writing to a temp location (would return a File)

cmnbroad commented:
@droazen please take a look at the updated README.md in this PR.

droazen commented on src/main/java/org/broadinstitute/hellbender/utils/io/IOUtils.java:
```diff
@@ -229,6 +229,20 @@ public static File writeTempResource(Resource resource) {
+    /**
+     * Create a resource from a path and a relative class, and write it to a temporary file.
+     * @param resourcePath Relative or absolute path to the class.
+     * @param relativeClass Relative class to use as a class loader and for a relative package.
```
Javadoc `@param` tags usually come at the end, rather than in the middle of the comment text.

cmnbroad replied:
Doh.

droazen commented on src/test/java/org/broadinstitute/hellbender/utils/io/IOUtilsUnitTest.java:
```diff
@@ -270,5 +267,42 @@ public void testUrlEncodeDecode(final String string, final String encoded) {
+        finally {
+            resourceFile.delete();
+        }
+        Assert.assertEquals(resourceContents, "line in large resource file");
```
Declare expected text as a constant, and use it in both test methods.

droazen commented on src/test/java/org/broadinstitute/hellbender/utils/io/IOUtilsUnitTest.java:
```diff
@@ -270,5 +267,42 @@ public void testUrlEncodeDecode(final String string, final String encoded) {
+    @DataProvider(name="resourcePaths")
+    public Object[][] getResourcePaths() {
+        return new Object[][] {
+                { Resource.LARGE_RUNTIME_RESOURCES_PATH + "/largeResourceTest.txt", null }
```
Can you add a test case involving a non-null relative class as well?

droazen commented on README.md:
```diff
@@ -50,28 +50,36 @@ releases of the toolkit.
+    * [git-lfs](https://git-lfs.github.com/) 1.1.0 or greater. Required to download the large files used to build GATK and
+      run the test suite. Run `git lfs install` after downloading, followed by `git lfs pull` from the root of your git clone
+      to download all of the large files. The full download is approximately 2 gigabytes. Alternatively, to pull the minimal
+      set of large `lfs` resource files required to only build ()without running tests), the test resources can be excluded
```
`()without running tests)` -> `(without running tests)`

cmnbroad replied:
Done.

cmnbroad replied:
Done.

droazen commented on README.md:
```diff
@@ -50,28 +50,36 @@ releases of the toolkit.
+      run the test suite. Run `git lfs install` after downloading, followed by `git lfs pull` from the root of your git clone
+      to download all of the large files. The full download is approximately 2 gigabytes. Alternatively, to pull the minimal
+      set of large `lfs` resource files required to only build ()without running tests), the test resources can be excluded
+      from the download by running `git lfs pull --include src/main/resources/large/` instead of `git lfs pull`. This greatly
```
Can we modify `build.gradle` to run this command automatically at build time, if lfs stubs are detected? Let's do that (if possible) -- then this set of instructions can be simplified to just "install git lfs"

cmnbroad replied:
Done. I changed the build to check for stubs, and to only run `git lfs pull --include src/main/resources/large/` if it finds any, then validate again. Also, we still have to manually download the files into the staging area, since the Docker doesn't have git-lfs installed.

cmnbroad replied:
Done see above note.

droazen commented on README.md:
```diff
@@ -50,28 +50,36 @@ releases of the toolkit.
-    * [git-lfs](https://git-lfs.github.com/) 1.1.0 or greater (needed to download large files for the complete test suite).
-      Run `git lfs install` after downloading, followed by `git lfs pull` from the root of your git clone to download the large files. The download is several hundred megabytes.
-* Alternatively, pre-packaged images with all needed dependencies installed can be found on [our dockerhub repository](https://hub.docker.com/r/broadinstitute/gatk/). This requires a recent version of the docker client, which can be found on the [docker website](https://www.docker.com/get-docker).
+* To build GATK:
```
Is R a build-time dependency as well?

cmnbroad replied:
Only to run tests, AFAIK. I added it to this section as well saying its needed for running tests.

cmnbroad commented on build.gradle:
```diff
@@ -66,8 +67,42 @@ final testNGVersion = '6.11'
+// Validate that we have a clone of the repository and that any required git-lfs files have been resolved
+def validateBuildPrerequisites(javaVersion, largeResourcesFolder, lfsSentinel, buildPrerequisitesMessage) {
+    if (ToolProvider.getSystemToolClassLoader() == null) {
+        throw new GradleException("The SystemToolClassLoader obtained from the ToolProvider is null. A Java $RequiredJavaVersion JDK must be installed. $buildPrerequisitesMessage")
```
I think it happens when you have a JRE rather than a JDK (see https://github.com/broadinstitute/gatk/issues/4532). Also, although I haven't tried it I suspect it will happen with JDK9.

cmnbroad commented on build.gradle:
```diff
@@ -66,8 +67,42 @@ final testNGVersion = '6.11'
-
 final pythonPackageArchiveName = 'gatkPythonPackageArchive.zip'
+final largeResourcesFolder = "src/main/resources/large"
+final lfsFileSentinel = "version https://git-lfs.github.com/spec/v1"
```
I changed all of these names to use "stub", but I'm not sure thats better, since AFACIT git-lfs refers to these as pointer files in all of the help and doc.

cmnbroad commented on build.gradle:
```diff
@@ -66,8 +67,42 @@ final testNGVersion = '6.11'
+    final byte[] actualBytes = readBytesFromFile(largeResourceFile, lfsSentinel.length());
+    if (new String(actualBytes, "UTF-8") == lfsSentinel) {
+        throw new GradleException(
+                "$largeResourceFile appears to be a git-lfs pointer file." +
```
Done.

cmnbroad commented on build.gradle:
```diff
@@ -66,8 +67,42 @@ final testNGVersion = '6.11'
+    if (new String(actualBytes, "UTF-8") == lfsSentinel) {
+        throw new GradleException(
+                "$largeResourceFile appears to be a git-lfs pointer file." +
+                        " The resource files in src/main/resources/large must be downloaded using git-lfs. " +
```
Done.

cmnbroad commented on build.gradle:
```diff
@@ -66,8 +67,42 @@ final testNGVersion = '6.11'
+final buildPrerequisitesMessage = "See https://github.com/broadinstitute/gatk#building for information on how to build GATK."
+
+// Large runtime resource files must be present at build time to be compiled into the jar, so validate
+// that the lfs files have been fetched and that the directory does not contain the git-lfs pointer files
```
Done.

cmnbroad commented on build.gradle:
```diff
@@ -66,8 +67,42 @@ final testNGVersion = '6.11'
+        throw new GradleException("The GATK Github repository must be cloned using \"git clone\" to run the build. $buildPrerequisitesMessage")
+    }
+    // Next, ensure that the git-lfs files required to run the build are present
+    def largeResources = fileTree(dir: largeResourcesFolder)
```
Only files.

cmnbroad commented:
Back to @droazen. Responded to all of the comments. I did an interim commit to test the lfs changes on docker/travis so there are two commits for this round.

droazen commented on src/test/java/org/broadinstitute/hellbender/utils/io/IOUtilsUnitTest.java:
```diff
@@ -270,5 +267,44 @@ public void testUrlEncodeDecode(final String string, final String encoded) {
+    {
+        final Resource largeResource = new Resource(resourcePath, relativeClass);
+        final File resourceFile = IOUtils.writeTempResource(largeResource);
+        final String resourceContentsFirstLine = getFirstLineAndDeleteTempFile(resourceFile);
```
Should the `IOUtils.writeTempResource()` methods set the extracted resource to be deleted on JVM exit? It seems like we will end up leaving garbage around if we don't do that in the methods themselves.

cmnbroad replied:
We probably should. I'm going to do that in a separate PR though since all of the call sites should be cleaned up as well.

droazen commented on README.md:
```diff
@@ -50,28 +50,37 @@ releases of the toolkit.
+    * [git-lfs](https://git-lfs.github.com/) 1.1.0 or greater. Required to download the large files used to build GATK, and
+      test files required to run the test suite. Run `git lfs install` after downloading, followed by `git lfs pull` from
+      the root of your git clone to download all of the large files, including those required to run the test suite. The
+      full download is approximately 2 gigabytes. Alternatively, you can skip this step since the build will itself use
```
"Alternatively, " -> "Alternatively, if you are just building GATK and not running the test suite, "

cmnbroad commented on build.gradle:
```diff
@@ -66,8 +67,66 @@ final testNGVersion = '6.11'
+        println "Fetching large runtime resources via: $execLFSPull"
+        def retCode = execLFSPull.execute().waitFor()
+        if (retCode.intValue() != 0) {
+            throw new GradleException("git-lfs pull execution failed: $retCode");
```
Done.

cmnbroad commented on build.gradle:
```diff
@@ -66,8 +67,66 @@ final testNGVersion = '6.11'
+        }
+        return retCode
+    } catch (IOException e) {
+        throw new GradleException("git-lfs must be installed to run the GATK build", e)
```
I can't be sure. I changed the message to report the IOException and suggest checking git-lfs.

cmnbroad commented on build.gradle:
```diff
@@ -66,8 +67,66 @@ final testNGVersion = '6.11'
+        if (checkLargeResourcesForStubFiles(largeResourcesFolder, lfsStubFileHeader, buildPrerequisitesMessage)) {
+            throw new GradleException(
+                    "$largeResourcesFolder contains git-lfs stub files." +
+                            " The resource files in $largeResourcesFolder must be downloaded using git-lfs. " +
```
Done.

cmnbroad commented on build.gradle:
```diff
@@ -66,8 +67,66 @@ final testNGVersion = '6.11'
+}
+
+// Validate that we have a clone of the repository and that any required git-lfs files have been resolved
+def validateBuildPrerequisites(javaVersion, largeResourcesFolder, lfsStubFileHeader, buildPrerequisitesMessage) {
```
done.

cmnbroad commented on build.gradle:
```diff
@@ -262,6 +321,8 @@ processResources {
     //Don't waste time packaging unnecessary test data into the test resources:
     include "org/broadinstitute/hellbender/utils/config/*"
+    //Required for IOUtils resource tests
+    include "org/broadinstitute/hellbender/utils/io/*"
```
Only if you add a resource that needs to be packaged into the test jar as part of the test resources.

droazen commented on README.md:
```diff
@@ -50,28 +50,38 @@ releases of the toolkit.
+      the root of your git clone to download all of the large files, including those required to run the test suite. The
+      full download is approximately 2 gigabytes. Alternatively, if you are just building GATK and not running the test
+      suite, you can skip this step since the build itself will use git-lfs to download the minimal set of large `lfs`
+      resource files required to complete the build. The tests resources will not be downloaded, but this greatly reduces
```
"tests resources" -> "test resources"

cmnbroad replied:
Done.

droazen commented on build.gradle:
```diff
@@ -68,6 +69,72 @@ final baseJarName = 'gatk'
+    final lfsStubFileHeader = "version https://git-lfs.github.com/spec/v1"  // first line of a git-lfs stub file
+    def readBytesFromFile = { largeFile, n ->
+        final byte[] bytes = new byte[n]
+        largeFile.withInputStream { stream -> stream.read(bytes, 0, bytes.length) }
```
Does the input stream here get closed?

cmnbroad replied:
Yes, its equivalent to try-wth-resources.
